src/io/video.js

```diff
@@ -0,0 +1,332 @@
+     */
+    getFrame ({
+        dimensions = Video.DIMENSIONS,
+        mirror = this.mirror,
+        format = Video.FORMAT_IMAGE_DATA,
+        cacheTimeout = this._frameCacheTimeout
+    }) {
+        if (!this.videoReady) {
+            return null;
+        }
+        const [width, height] = dimensions;
+        const workspace = this._getWorkspace({dimensions, mirror: Boolean(mirror)});
+        const {videoWidth, videoHeight} = this._video;
+        const {canvas, context, lastUpdate, cacheData} = workspace;
+        const now = Date.now();
+
+        // if the canvas hasn't been updated...
+        if (lastUpdate + cacheTimeout < now) {
+
+            if (mirror) {
+                context.scale(-1, 1);
+                context.translate(width * -1, 0);
+            }
+
+            context.drawImage(this._video,
+                // source x, y, width, height
+                0, 0, videoWidth, videoHeight,
+                // dest x, y, width, height
+                0, 0, width, height
+            );
+
+            context.resetTransform();
+            workspace.lastUpdate = now;
+        }
+
+        // each data type has it's own data cache, but the canvas is the same
+        if (!cacheData[format]) {
+            cacheData[format] = {lastUpdate: 0};
+        }
+        const formatCache = cacheData[format];
+
+        if (formatCache.lastUpdate + cacheTimeout < now) {
+            if (format === Video.FORMAT_IMAGE_DATA) {
+                formatCache.lastData = context.getImageData(0, 0, width, height);
+            } else if (format === Video.FORMAT_CANVAS) {
+                // this will never change
+                formatCache.lastUpdate = Infinity;
+                formatCache.lastData = canvas;
+            } else {
+                log.error(`video io error - unimplemented format ${format}`);
+                // cache the null result forever, don't log about it again..
+                formatCache.lastUpdate = Infinity;
+                formatCache.lastData = null;
+            }
+
+            // rather than set to now, this data is as stale as it's canvas is
+            formatCache.lastUpdate = Math.max(workspace.lastUpdate, formatCache.lastUpdate);
+        }
+
+        return formatCache.lastData;
+    }
+
+    /**
+     * Set the preview ghost effect
+     * @param {number} ghost from 0 (visible) to 100 (invisible) - ghost effect
+     */
+    setPreviewGhost (ghost) {
+        if (this._drawable) {
+            this.runtime.renderer.updateDrawableProperties(this._drawable, {ghost});
+        }
+    }
+
+
+    /**
+     * Create a video stream.
+     * Should probably be moved to -render or somewhere similar later
+     * @private
+     * @return {Promise} When video has been received, rejected if video is not received
+     */
+    _setupVideo () {
+        if (this._singleSetup) {
+            return this._singleSetup;
+        }
+
+        this._video = document.createElement('video');
```

Comment:
There is a race condition on this member `this._video`. It affects a lot of the code in here relating to enabling and disabling the video stream. There are a few ways we can fix it.

The first I can think of is to put all the enabling and disabling code into a task queue. We could use a separate promise chain from `_singleSetup` to create that queue. That queue would also help manage the race conditions in the video preview.

Another fix, would be the more thorough stage change tracking so the right parts of the video setup calls execute or don't. (Internally there is more than if the video is on or off.)

-----

An extra credit change would be to set a state value to enabled or disabled and debounce the enabling/disabling work. When the debounce'd work starts, check the enabled or disabled state and kick off that work. That way quick camera toggling doesn't create multiple video tags or getUserMedia calls.

Reply:
What is the race you're seeing? enable/disable/enable causing two getUserMedia calls to overwrite each other's stream?  I think the only race-blocking needed would be stopping the old `_track` if we had one.

Reply:
I did a bunch of extra checks to make sure the state was what I expected...

I also tested in my gui with some blocks like 

![image 1](https://user-images.githubusercontent.com/549355/38381843-0e9ef5e6-38d6-11e8-8a65-300e792bb129.png)

It seems to do the "right thing" which is end in whatever state you asked for.

Reply:
^^^ in this state there might end up being like 3 or 4 getUserMedia calls going at once, but they all clean themselves up and only one stream is kept

Reply:
The race condition here involved the video tag and preview work without the checks that have since been added.

Reply:
Wish I could mark this as resolved.